### PR TITLE
feat: blob reporter

### PR DESCRIPTION
**Pull request by yury-s**
#10437

### CODE REVIEW

File: packages/playwright-test/src/reporters/blob.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+export class BlobReporter extends TeleReporterEmitter {
+  private _messages: any[] = [];
+  private _options: BlobReporterOptions;
+  private _outputFile: string | undefined;
+
+  constructor(options: BlobReporterOptions = {}) {
+    super(message => this._messages.push(message));
+    this._options = options;
+  }
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/blob.ts:42]: lint blank line below.

File: packages/playwright-test/src/reporters/blob.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  override onBegin(config: FullConfig<{}, {}>, suite: Suite): void {
+    super.onBegin(config, suite);
+    const outputDir = this._resolveOutputDir(config as FullConfigInternal);
+    let shardSuffix = '';
+    if (config.shard) {
+      const paddedNumber = String(config.shard.current).padStart(String(config.shard.total).length, '0');
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/blob.ts:48]: ChatGPT is suggesting

```js
const paddedNumber = `${config.shard.current}`.padStart(`${config.shard.total}`.length, '0');
```

File: packages/playwright-test/src/reporters/blob.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  private _options: BlobReporterOptions;
+  private _outputFile: string | undefined;
+
+  constructor(options: BlobReporterOptions = {}) {
+    super(message => this._messages.push(message));
+    this._options = options;
+  }
+  override onBegin(config: FullConfig<{}, {}>, suite: Suite): void {
+    super.onBegin(config, suite);
+    const outputDir = this._resolveOutputDir(config as FullConfigInternal);
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/blob.ts:45]: Extract output file calculation into a separate method, inline resolveOutputDir there.

File: packages/playwright-test/src/reporters/blob.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    if (config.shard) {
+      const paddedNumber = String(config.shard.current).padStart(String(config.shard.total).length, '0');
+      shardSuffix = `-${paddedNumber}-of-${config.shard.total}`;
+    }
+    this._outputFile = path.join(outputDir, `report${shardSuffix}.zip`);
+  }
+
+  override async onEnd(result: FullResult): Promise<void> {
+    await super.onEnd(result);
+    if (!this._outputFile)
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/blob.ts:56]: How can this happen? Looks like we should have thrown when calculating the output file?

File: packages/playwright-test/src/reporters/blob.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+      return;
+
+    const reportString = JSON.stringify(this._messages, undefined, 2);
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/blob.ts:59]: Let's use the JSONL format for serialization / parsing.

File: packages/playwright-test/src/reporters/blob.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    fs.mkdirSync(path.dirname(this._outputFile), { recursive: true });
+    await zipReport(this._outputFile, reportString);
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/blob.ts:61]: ... and stream the messages into the zip, preferably live.

File: packages/playwright-test/src/reporters/blob.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+export async function createMergedReport(config: FullConfigInternal, dir: string, reporterName?: string) {
+  const shardFiles = await sortedShardFiles(dir);
+  console.log(`Merging ${shardFiles}`);
+  const events = await mergeEvents(dir, shardFiles);
+  console.log(`Writing ${reporterName} report to ${dir}`);
+
+  // TODO: use reporterName.
+  const reporter = new HtmlReporter({ _internalResolvedOutputFolder: dir });
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/blob.ts:79]: Looks like you should implement this todo right away

File: packages/playwright-test/src/reporters/blob.ts [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  const receiver = new TeleReporterReceiver(path.sep, reporter);
+  // TODO: handle onError.
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/blob.ts:81]: Ditto

File: packages/playwright-test/src/reporters/blob.ts [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+  return {
+    method: 'onBegin',
+    params: {
+      config: beginEvents[0].params.config,
+      projects,
+    }
+  };
+}
+
+function mergeJsonSuites(jsonSuites: JsonSuite[], parent: JsonSuite | JsonProject) {
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/blob.ts:138]: There is similar code in the telereceiver, what is the difference.

File: packages/playwright-test/src/reporters/html.ts [hunk 10]
```diff
@@ -42,6 +42,7 @@ const kMissingContentType = 'x-playwright/missing';
 type HtmlReportOpenOption = 'always' | 'never' | 'on-failure';
 type HtmlReporterOptions = {
   outputFolder?: string,
+  _internalResolvedOutputFolder?: string,
```
Comment (pavelfeldman) [inline packages/playwright-test/src/reporters/html.ts:45]: Why can't outputFolder be resolved?

File: packages/playwright-test/src/reporters/blob.ts (hunk 9 above)
Comment (yury-s) [reply packages/playwright-test/src/reporters/blob.ts:138]: This one operates on Json* structures while that works with Tele* and the logic is different in many ways.

File: packages/playwright-test/src/reporters/blob.ts (hunk 6 above)
Comment (yury-s) [reply packages/playwright-test/src/reporters/blob.ts]: Let's add it separately.

**Review by pavelfeldman — APPROVED** [review]

**Pull request merged**